File: src/main/java/org/gecko/io/ViewModelElementSaver.java
package org.gecko.io;

import java.util.ArrayList;
import java.util.List;
import javafx.scene.paint.Color;
import org.gecko.model.Automaton;
import org.gecko.model.Edge;
import org.gecko.model.Region;
import org.gecko.model.State;
import org.gecko.model.System;
import org.gecko.model.SystemConnection;
import org.gecko.model.Variable;
import org.gecko.viewmodel.GeckoViewModel;
import org.gecko.viewmodel.PositionableViewModelElement;
import org.gecko.viewmodel.RegionViewModel;

/**
 * Performs operations for every {@link org.gecko.model.Element Model-Element} from the subtree of a
 * {@link System}, creating for each of them a {@link ViewModelPropertiesContainer}, depending on the attributes of the
 * corresponding {@link PositionableViewModelElement}.
 */
public class ViewModelElementSaver {
    private final GeckoViewModel geckoViewModel;
    private final List<ViewModelPropertiesContainer> viewModelProperties;

    protected ViewModelElementSaver(GeckoViewModel geckoViewModel) {
        this.geckoViewModel = geckoViewModel;
        this.viewModelProperties = new ArrayList<>();
    }

    public void saveStateViewModelProperties(State state) {
        ViewModelPropertiesContainer stateViewModelContainer =
            this.getCoordinateContainer(this.geckoViewModel.getViewModelElement(state));
        this.viewModelProperties.add(stateViewModelContainer);
    }

    public void saveRegionViewModelProperties(Region region) {
        ViewModelPropertiesContainer regionViewModelContainer =
            this.getCoordinateContainer(this.geckoViewModel.getViewModelElement(region));

        Color color = ((RegionViewModel) this.geckoViewModel.getViewModelElement(region)).getColor();
        regionViewModelContainer.setRed(color.getRed());
        regionViewModelContainer.setGreen(color.getGreen());
        regionViewModelContainer.setBlue(color.getBlue());

        this.viewModelProperties.add(regionViewModelContainer);
    }

    public void saveSystemViewModelProperties(System system) {
        ViewModelPropertiesContainer systemViewModelContainer =
            this.getCoordinateContainer(this.geckoViewModel.getViewModelElement(system));
        this.viewModelProperties.add(systemViewModelContainer);
    }

    public void saveSystemConnectionViewModelProperties(SystemConnection systemConnection) {
        ViewModelPropertiesContainer systemConnectionViewModelContainer =
            this.getCoordinateContainer(this.geckoViewModel.getViewModelElement(systemConnection));
        this.viewModelProperties.add(systemConnectionViewModelContainer);
    }

    public void saveEdgeModelProperties(Edge edge) {
        ViewModelPropertiesContainer edgeViewModelContainer =
            this.getCoordinateContainer(this.geckoViewModel.getViewModelElement(edge));
        this.viewModelProperties.add(edgeViewModelContainer);
    }

    public void savePortViewModelProperties(Variable variable) {
        ViewModelPropertiesContainer variableViewModelContainer =
            this.getCoordinateContainer(this.geckoViewModel.getViewModelElement(variable));
        this.viewModelProperties.add(variableViewModelContainer);
    }

    private ViewModelPropertiesContainer getCoordinateContainer(PositionableViewModelElement<?> element) {
        ViewModelPropertiesContainer container = new ViewModelPropertiesContainer();
        container.setElementId(element.getTarget().getId());
        container.setId(element.getId());
        container.setPositionX(element.getPosition().getX());
        container.setPositionY(element.getPosition().getY());
        container.setSizeX(element.getSize().getX());
        container.setSizeY(element.getSize().getY());
        return container;
    }

    protected List<ViewModelPropertiesContainer> getViewModelProperties(System root) {
        this.gatherSystemAttributes(root);
        return this.viewModelProperties;
    }

    private void gatherSystemAttributes(System system) {
        for (Variable variable : system.getVariables()) {
            this.savePortViewModelProperties(variable);
        }

        for (SystemConnection systemConnection : system.getConnections()) {
            this.saveSystemConnectionViewModelProperties(systemConnection);
        }

        Automaton automaton = system.getAutomaton();

        for (Region region : automaton.getRegions()) {
            this.saveRegionViewModelProperties(region);
        }

        for (State state : automaton.getStates()) {
            this.saveStateViewModelProperties(state);
        }

        for (Edge edge : automaton.getEdges()) {
            this.saveEdgeModelProperties(edge);
        }

        for (System child : system.getChildren()) {
            this.saveSystemViewModelProperties(child);
            this.gatherSystemAttributes(child);
        }
    }
}

File: src/main/kotlin/org/gecko/io/ViewModelElementSaver.kt
package org.gecko.io


import org.gecko.viewmodel.*

/**
 * Performs operations for every [Model-Element][org.gecko.model.Element] from the subtree of a [System],
 * creating for each of them a [ViewModelPropertiesContainer], depending on the attributes of the corresponding
 * [PositionableElement].
 */
class ViewModelElementSaver(val gModel: GModel) {
    val viewModelProperties: MutableList<ViewModelPropertiesContainer> = arrayListOf()
    val startStates: MutableList<StartStateContainer> = arrayListOf()

    fun getViewModelProperties(root: System): List<ViewModelPropertiesContainer> {
        gatherSystemAttributes(root)
        return this.viewModelProperties
    }

    fun gatherSystemAttributes(system: System) {
        for (variable in system.ports) {
            this.savePortViewModelProperties(variable)
        }

        for (systemConnection in system.connections) {
            this.saveSystemConnectionViewModelProperties(systemConnection)
        }

        val automaton = system.automaton

        /*if (automaton.startState != null) {
            val startStateContainer = StartStateContainer()
            //startStateContainer.systemId = system.hashCode()
            startStateContainer.startStateName = automaton.startState!!.name
            startStates.add(startStateContainer)
        }*/

        for (region in automaton.regions) {
            this.saveRegionViewModelProperties(region)
        }

        for (state in automaton.states) {
            this.saveStateViewModelProperties(state)
        }

        for (edge in automaton.edges) {
            this.saveEdgeModelProperties(edge)
        }

        for (child in system.subSystems) {
            this.saveSystemViewModelProperties(child)
            this.gatherSystemAttributes(child)
        }
    }

    fun saveStateViewModelProperties(state: State) {
        val stateViewModelContainer = this.getCoordinateContainer(state)
        viewModelProperties.add(stateViewModelContainer)
    }

    fun saveRegionViewModelProperties(region: Region) {
        val regionViewModelContainer = this.getCoordinateContainer(region)

        val color = region.color
        regionViewModelContainer.red = color.red
        regionViewModelContainer.green = color.green
        regionViewModelContainer.blue = color.blue

        viewModelProperties.add(regionViewModelContainer)
    }

    fun saveSystemViewModelProperties(system: System) {
        val systemViewModelContainer = this.getCoordinateContainer((system))
        viewModelProperties.add(systemViewModelContainer)
    }

    fun saveSystemConnectionViewModelProperties(systemConnection: SystemConnection) {
        val systemConnectionViewModelContainer = this.getCoordinateContainer((systemConnection))
        viewModelProperties.add(systemConnectionViewModelContainer)
    }

    fun saveEdgeModelProperties(edge: Edge) {
        val edgeViewModelContainer = this.getCoordinateContainer(edge)
        viewModelProperties.add(edgeViewModelContainer)
    }

    fun savePortViewModelProperties(variable: Port) {
        val variableViewModelContainer = this.getCoordinateContainer(variable)
        viewModelProperties.add(variableViewModelContainer)
    }

    fun getCoordinateContainer(element: PositionableElement): ViewModelPropertiesContainer {
        val container = ViewModelPropertiesContainer()
        //container.elementId = element.hashCode()
        container.id = element.hashCode()
        container.positionX = element.position.x
        container.positionY = element.position.y
        container.sizeX = element.size.x
        container.sizeY = element.size.y
        return container
    }
}
